src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -0,0 +1,283 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import itertools
+import logging
+import os
+from pathlib import Path
+from typing import Any, Iterable, Mapping, Optional
+
+import toml
+from packaging.version import InvalidVersion, Version
+from pkg_resources import Requirement
+
+from pants.base.build_environment import get_buildroot
+
+logger = logging.getLogger(__name__)
+
+
+def get_max_caret(proj_name: str, version: str, req: str, fp: str) -> str:
+    major = 0
+    minor = 0
+    micro = 0
+
+    try:
+        parsed_version = Version(version)
+    except InvalidVersion:
+        raise InvalidVersion(
+            f"Failed to parse requirement {req} in {fp} loaded by the poetry_requirements macro.\n\nIf you believe this requirement is valid, consider opening an issue at https://github.com/pantsbuild/pants/issues so that we can update Pants's Poetry macro to support this."
+        )
+
+    if parsed_version.major != 0:
+        major = parsed_version.major + 1
+    elif parsed_version.minor != 0:
+        minor = parsed_version.minor + 1
+    elif parsed_version.micro != 0:
+        micro = parsed_version.micro + 1
+    else:
+        base_len = len(parsed_version.base_version.split("."))
+        if base_len >= 3:
+            micro = 1
+        elif base_len == 2:
+            minor = 1
+        elif base_len == 1:
+            major = 1
+
+    return f"{major}.{minor}.{micro}"
+
+
+def get_max_tilde(proj_name: str, version: str, req: str, fp: str) -> str:
+    major = 0
+    minor = 0
+    try:
+        parsed_version = Version(version)
+    except InvalidVersion:
+        raise InvalidVersion(
+            f'Failed to parse requirement {proj_name} = "{req}" in {fp} loaded by the poetry_requirements macro.\n\nIf you believe this requirement is valid, consider opening an issue at https://github.com/pantsbuild/pants/issues so that we can update Pants\'s Poetry macro to support this.'
+        )
+    base_len = len(parsed_version.base_version.split("."))
+    if base_len >= 2:
+        minor = parsed_version.minor + 1
+        major = parsed_version.major
+    elif base_len == 1:
+        major = parsed_version.major + 1
+
+    return f"{major}.{minor}.0"
+
+
+def parse_str_version(proj_name: str, attributes: str, fp: str) -> str:
+    valid_specifiers = "<>!~="
+    pep440_reqs = []
+    comma_split_reqs = (i.strip() for i in attributes.split(","))
+    for req in comma_split_reqs:
+        if req[0] == "^":
+            max_ver = get_max_caret(proj_name, req[1:], req, fp)
```

Comment:
I've been trying to think about how to remove the duplication w/ parsing the `Version` and re-raising a new exception. It's not the End Of The World to duplicate it, but not ideal, e.g. if you change the error message and forget to update the other.

One way you could de-duplicate is like this:

```python
is_caret = req[0] and "^"
# ~= is an acceptable default operator; however, ~ is not, and IS NOT the same as ~=
is_tilde = req[0] == "~" and req[1] != "="
if is_caret or is_tilde:
     try:
          version = Version(req[1])
     except ..:
          ...
     max_ver = get_max_caret(version) if is_caret else get_max_tilde(version)
     min_ver = req[1:]
     pep440_reqs.append(f">={min_ver},<{max_ver}")
else:
    pep440_reqs.append(req if req[0] in valid_specifiers else f"=={req}")
```

Note that this allows you to simplify the `get_max_tilde` and `get_max_caret` functions to take fewer arguments.